cpp/oneapi/dal/algo/triangle_counting/detail/relabel_kernel.hpp

```diff
@@ -0,0 +1,120 @@
+    std::int32_t* new_ids,
+    std::int32_t* degrees_relabel,
+    std::int64_t vertex_count);
+
+ONEDAL_EXPORT void parallel_prefix_sum(const dal::detail::host_policy& policy,
+                                       const std::int32_t* degrees_relabel,
+                                       std::int64_t* offsets,
+                                       std::int64_t* part_prefix,
+                                       std::int64_t* local_sums,
+                                       std::int64_t block_size,
+                                       std::int64_t num_blocks,
+                                       std::int64_t vertex_count);
+
+ONEDAL_EXPORT void fill_relabeled_topology(const dal::detail::host_policy& policy,
+                                           const dal::preview::detail::topology<std::int32_t>& t,
+                                           std::int32_t* vertex_neighbors_relabel,
+                                           std::int64_t* edge_offsets_relabel,
+                                           std::int64_t* offsets,
+                                           const std::int32_t* new_ids);
+
+template <typename Allocator>
+struct relabel_by_greater_degree {
+    inline void operator()(const dal::detail::host_policy& ctx,
+                           const dal::preview::detail::topology<std::int32_t>& t,
+                           std::int32_t* vertex_neighbors_relabel,
+                           std::int64_t* edge_offsets_relabel,
+                           std::int32_t* degrees_relabel,
+                           const Allocator& alloc) {
+        const auto vertex_count = t.get_vertex_count();
+        using int32_allocator_type =
+            typename std::allocator_traits<Allocator>::template rebind_alloc<std::int32_t>;
+
+        using pair_allocator_type = typename std::allocator_traits<
+            Allocator>::template rebind_alloc<std::pair<std::int32_t, std::size_t>>;
+
+        using int64_allocator_type =
+            typename std::allocator_traits<Allocator>::template rebind_alloc<std::int64_t>;
+
+        int64_allocator_type int64_allocator(alloc);
+        pair_allocator_type pair_allocator(alloc);
+        int32_allocator_type int32_allocator(alloc);
+
+        std::pair<std::int32_t, std::size_t>* degree_id_pairs =
+
+            oneapi::dal::preview::detail::allocate(pair_allocator, vertex_count);
+
+        sort_ids_by_degree(ctx, t._degrees.get_data(), degree_id_pairs, vertex_count);
+
+        std::int32_t* new_ids =
+            oneapi::dal::preview::detail::allocate(int32_allocator, vertex_count);
+
+        fill_new_degrees_and_ids(ctx, degree_id_pairs, new_ids, degrees_relabel, vertex_count);
+
+        oneapi::dal::preview::detail::deallocate(pair_allocator, degree_id_pairs, vertex_count);
+
+        std::int64_t* offsets =
+            oneapi::dal::preview::detail::allocate(int64_allocator, vertex_count + 1);
+
+        const size_t block_size = 1 << 20;
```

Comment:
why size_t
1<<20 -  a magic const. Would it depend on cpu?

Reply:
Size_t again was copied from the prototype.
Magic const value is near 1 million. It was empirically found at the prototype level. Its purpose is O(1)  heuristic to choose block size while computing parallel prefix sum aka CSR "rows".